protocol/README.md

```diff
@@ -4,38 +4,40 @@ This document describes the canonical system architecture used to encrypt and de
+    * Read the TDF Client public key from a custom HTTP header sent with the OIDC authentication request.
+    * Construct and send an Attribute Provider web service request, including the public key in the payload.
+    * Return the resulting Claims Object in the signed IdP JWT.
+  * *Virtru Protocol Mapper* is our Keycloak-specific reference implementation of the above functionality.
+* *Attribute Provider* (AP) - A web service that will receive requests from our OIDC IdP custom code (ex: Virtru Protocol Mapper) containing information about authenticated users and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).
+* *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext.  A valid OIDC token containing [TDF Claims](../schema/ClaimsObject.md) must be used as a bearer token when communicating with KAS. An otherwise valid and trusted OIDC token without valid TDF Claims will be rejected as a bearer token by KAS.
 
 ## Workflow
 
 The following sequence diagrams illustrate the workflow taken by the client to encrypt or decrypt TDF ciphertext. The canonical TDF architecture supports two modes of operation: _online mode_ and _offline mode_, which have distinct workflows as shown below.
 
 _Online mode_ is the default mode, where the [wrapped data key](../schema/KeyAccessObject.md) and [authorization policy](../schema/PolicyObject.md) for TDF ciphertext is committed to KAS in-band as part of the `encrypt` operation. This means that the `encrypt` will succeed if and only if all resources are prepared to facilitate an immediate decrypt.
 
-_Offline mode_ requires that the TDF Client previously authenticated with EAS to get a valid Entity Object. Clients running in this mode commit the [authorization policy](../schema/PolicyObject.md) out-of-band, or when decrypt is first performed. This significantly reduces latency at the cost of larger ciphertext (as the entire wrapped key and policy information must be included). 
-
-Long-lived entity objects can be made available for certain _offline mode_ use cases, for instance, sensors with intermittent or low-bandwidth connectivity can have an Entity Object flashed onto the device during manufacturing.
+_Offline mode_ requires that the TDF Client previously authenticated with OIDC to get a valid Claims Object. Clients running in this mode commit the [authorization policy](../schema/PolicyObject.md) out-of-band, or when decrypt is first performed. This significantly reduces latency at the cost of larger ciphertext (as the entire wrapped key and policy information must be included). 
```

Comment:
The cipher text isn't larger. But there is a slightly larger manifest bundled with the TDF for the reasons mentioned here.

Also, I'm not sure that the client must have authenticated to get the claims object, unless something has changed that I'm not aware of.

Reply:
To be clear, what I mean is that I'm not sure a claims object is required for offline encrypt. I didn't mean to suggest that one needn't be authenticated to get a claims object - something that is obviously false.

Reply:
@biscoe916 At least for the preexisting JS SDK (`tdf3-js`), a network call to EAS to fetch a claims object is made unconditionally on every `encrypt` regardless of offline or online mode toggle - so that must have changed at some point.

The nanotdf and C++ SDKs support `setEntityObject` and will use the claims from a cached EO if present.

Currently the OIDC implementation **does** require the client to auth to get the claims object - in fact the client doesn't handle "claims objects" at all, it simply forwards the entire auth token it gets to KAS and doesn't have opinions about the content, or extract and parse any of it, by design - the bearer token and the client claims are one and the same.

The TDF3 clients however can simply be seeded with a JWT (previously obtained, containing claims) directly, which they can use as a KAS bearer token - this mimics the old `setEO`/offline claims model AFAICT.

Reply:
```
Long-lived bearer tokens can be made available for certain _offline mode_ use cases, for instance, sensors with intermittent or low-bandwidth connectivity can have an Entity Object flashed onto the device during manufacturing.
```

Obviously doing this with a JWT (or any clientside auth token/secret/credential/etc) is a terrible idea, so we might need to rethink this a bit if it's critical to support this flow with OIDC.